wallet/db.c

```diff
@@ -454,6 +460,86 @@ bool sqlite3_column_short_channel_id(sqlite3_stmt *stmt, int col,
 	size_t sourcelen = sqlite3_column_bytes(stmt, col);
 	return short_channel_id_from_str(source, sourcelen, dest);
 }
+bool sqlite3_bind_short_channel_id_array(sqlite3_stmt *stmt, int col,
+					 const struct short_channel_id *id)
+{
+	const tal_t *tmpctx;
+	char *allser;
+	char *ser;
+	size_t num;
+	size_t i;
+	bool res;
+
+	/* Handle nulls early. */
+	if (!id) {
+		sqlite3_bind_null(stmt, col);
+		return true;
+	}
+
+	tmpctx = tal_tmpctx(id);
+	allser = tal_strdup(tmpctx, "");
+	num = tal_count(id);
+
+	for (i = 0; i < num; ++i) {
+		ser = short_channel_id_to_str(tmpctx, id);
+		if (i == 0)
+			res = tal_append_fmt(&allser, "%s", ser);
+		else
+			res = tal_append_fmt(&allser, ",%s", ser);
+		if (!res) {
+			tal_free(tmpctx);
+			return false;
+		}
+	}
+
+	sqlite3_bind_blob(stmt, col, allser, strlen(allser), SQLITE_TRANSIENT);
+
+	tal_free(tmpctx);
+	return true;
+}
+struct short_channel_id *
+sqlite3_column_short_channel_id_array(const tal_t *ctx,
```

Comment:
We could simplify this a lot by using the `uint64` serialization format, so that each `short_channel_id` is exactly 8 bytes in size. No need for separator. That'd be a bit inconsistent with the way we serialize the `scid` in the `channels` table, but much simpler here.

Reply:
I think one concern is also byte order?  That seems to be the main reason for why scid is serialized as string rather than directly as blob.

Reply:
The byte order is fixed in the protocol, otherwise we couldn't communicate them at all. I just used the string serialization because I had it handy at the time, and we were stringifying everything else as well. We have since migrated to the `sqlite3_bind_*` interfaces for binaries, which are much safer and more performant.

Reply:
I suppose I was confused by what was meant by "`uint64` serialization format", are you referring to `fromwire_short_channel_id` and `towire_short_channel_id`?  If so, probably I should indeed use that.

It seems I can chain multiple calls to `towire_short_channel_id`, so maybe:

    for (i = 0; i < num; ++i) {
        towire_short_channel_id(&allser, &id[i]);
    }

Reply:
Yep, that's it :smiley: